cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkInstitutionLoginHandler.java

```diff
@@ -68,10 +69,33 @@ public Event getInstitutions(final RequestContext context) {
             throw new AssertionError("UTF-8 is unknown");
         }
 
-        final Map<String, String> institutions = this.institutionHandler.getInstitutionLoginUrls(target);
-        institutions.put("", " -- select an institution -- ");
-        final Map<String, String> sortedInstitutions = sortByValue(institutions);
-        context.getFlowScope().put("institutions", sortedInstitutions);
+        // Retrieve institution ID from flow context instead of URL params
+        String institutionId = null;
+        final String loginContext = (String) context.getFlowScope().get("jsonLoginContext");
+        final OpenScienceFrameworkLoginHandler.OpenScienceFrameworkLoginContext osfLoginContext
+                = OpenScienceFrameworkLoginHandler.OpenScienceFrameworkLoginContext.fromJson(loginContext);
+        if (osfLoginContext != null) {
+            institutionId = osfLoginContext.getInstitutionId();
+            // Set institution ID to null if not found
+            if (!this.institutionHandler.validateInstitutionForLogin(institutionId)) {
```

Comment:
This newly added check takes care of a corner case where the pre-selected institution is invalid and there is only one institution in total. In addition, this early check condenses extra condition checks when building the sorted map since `institutionId == null` after this check guarantees the all institutions path.